### ISSUE TITLE
Bluetooth support

### ISSUE THREAD

**Issue by sonicfelipemt** [Mxater/DualSenseSupport#5]
It only works wired, right? 
BLUTOOTH SOPPORT PLEAX

**Comment by SgtBreadStick** [comment]
Bluetooth wouldn't work as only 2 audio channels can be sent to the controller. The DualSense controller uses 4 audio channels for haptic and adaptive triggers. 2 channels for left/right haptics, and 2 channels for left/right triggers. Unless bluetooth gets changed to support more than 2 audio channels and the controller gets updated for it, it aint gonna happen.

However if Sony or someone released a USB dongle receiver which uses what the PS5 uses, then it can be done. So wired is the only option for now.

**Comment by sonicfelipemt** [comment]
Hm... Normal bluetooth adapters can't send more than 2 channels at the same time? Cuz Sony has standard BT 5.1 on the PS5. They must have their own way for sending more than 2 channels then...

**Comment by webwalkerkarma** [comment]
So. waiting when sony will release their official dongle I got another dumb idea that I will not be able to do myself. What if we will not use the sony haptic/sound solution and create our own. We probably be able to save 2 audio channels. I know it can be dumb and possibly time-consuming. I also don't know much about the level of control over the motors that available for us. But if it works, it (haptic solution) will be also applicable for other gamepads. So i suggest to take the audio stream, analyze what motors to use. discrete the hell out of the frequency and volume level value for each motor in use, and streamline the vibration strength proportional to volume level to the gamepad. We will also need a sound monitor function before the discretization processes.

**Comment by Etaash-mathamsetty** [comment]
hey guys, use my project if you want bluetooth support (I might consider making a windows build)
https://github.com/Etaash-mathamsetty/trigger-control

**Comment by Etaash-mathamsetty** [comment]
> Bluetooth wouldn't work as only 2 audio channels can be sent to the controller. The DualSense controller uses 4 audio channels for haptic and adaptive triggers. 2 channels for left/right haptics, and 2 channels for left/right triggers. Unless bluetooth gets changed to support more than 2 audio channels and the controller gets updated for it, it aint gonna happen.
> 
> However if Sony or someone released a USB dongle receiver which uses what the PS5 uses, then it can be done. So wired is the only option for now.

this is not true, (at least I don't think it's true), the triggers are actually controlled using regular hid data. The haptics are probably sound based or smth like that. 
the funny thing is that bluetooth support for adaptive triggers has been sitting under our noses the whole time, ds4windows has it implemented but in a pretty locked down, terrible way, luckily my project does give you full controller over the triggers over bluetooth. (i swear this isn't an ad, but im kinda just frustrated that nobody has made anything to get working wirelessly with all the control that this project gives)

**Comment by sonicfelipemt** [comment]
> this is not true, (at least I don't think it's true), the triggers are actually controlled using regular hid data. The haptics are probably sound based or smth like that. the funny thing is that bluetooth support for adaptive triggers has been sitting under our noses the whole time, ds4windows has it implemented but in a pretty locked down, terrible way, luckily my project does give you full controller over the triggers over bluetooth. (i swear this isn't an ad, but im kinda just frustrated that nobody has made anything to get working wirelessly with all the control that this project gives)

I took a look, and the controller on PS5 has a data stream for the adap triggers (not much different from the way it sends data for the touchpad and analogs), 2 audio channels (RearR + RearL) for vibration "haptics" and 2 audio channels for the 3.5mm audio jack output. 
That said, DS4windows even has emulation that converts regular dual-motor vibration data (Xbox / PS2~4 type) into the data type that works on the DualSense's vibrators. **And it even works via bluetoth!** That means the dev from that software cracked the stuff of sending audio via bluetooth. Or at least partially.

**Comment by Etaash-mathamsetty** [comment]
> > this is not true, (at least I don't think it's true), the triggers are actually controlled using regular hid data. The haptics are probably sound based or smth like that. the funny thing is that bluetooth support for adaptive triggers has been sitting under our noses the whole time, ds4windows has it implemented but in a pretty locked down, terrible way, luckily my project does give you full controller over the triggers over bluetooth. (i swear this isn't an ad, but im kinda just frustrated that nobody has made anything to get working wirelessly with all the control that this project gives)
> 
> I took a look, and the controller on PS5 has a data stream for the adap triggers (not much different from the way it sends data for the touchpad and analogs), 2 audio channels (RearR + RearL) for vibration "haptics" and 2 audio channels for the 3.5mm audio jack output. That said, DS4windows even has emulation that converts regular dual-motor vibration data (Xbox / PS2~4 type) into the data type that works on the DualSense's vibrators. **And it even works via bluetoth!** That means the dev from that software cracked the stuff of sending audio via bluetooth. Or at least partially.

I think sony's linux driver did that first, but yes that's how it works
another thing I find frustrating is that sony didn't bother with adaptive trigger support in the offical linux driver, I can hack it in, in 20 seconds time (with bluetooth support), so I wonder why sony can't do it...

**Comment by Etaash-mathamsetty** [comment]
ok so the windows build is ready if you guys wanna take it for a spin

**Comment by SgtBreadStick** [comment]
> ok so the windows build is ready if you guys wanna take it for a spin

More than happy to
